test/ZZStRSR.test.ts

```diff
@@ -2145,6 +2145,123 @@ describe(`StRSRP${IMPLEMENTATION} contract`, () => {
+      expect(await stRSR.exchangeRate()).to.be.closeTo(expectedRate, 1)
+
+      // Now governance can reset stakes
+      await expect(stRSR.connect(owner).resetStakes()).to.emit(stRSR, 'AllBalancesReset')
+
+      // All stakes reset
+      expect(await stRSR.exchangeRate()).to.equal(fp('1'))
+      expect(await stRSR.totalSupply()).to.equal(bn(0))
+      expect(await stRSR.balanceOf(addr1.address)).to.equal(bn(0))
+    })
+
+    it('Should reset stakes and perform validations on rate - MIN', async () => {
+      const stakeAmt: BigNumber = bn('1000e18')
+      const addAmt1: BigNumber = bn('100e18')
+      const addAmt2: BigNumber = bn('10e30')
+
+      // Stake
+      await rsr.connect(addr1).approve(stRSR.address, stakeAmt)
+      await stRSR.connect(addr1).stake(stakeAmt)
+
+      expect(await stRSR.exchangeRate()).to.equal(fp('1'))
+      expect(await stRSR.totalSupply()).to.equal(stakeAmt)
+      expect(await stRSR.balanceOf(addr1.address)).to.equal(stakeAmt)
+
+      // Cannot reset stakes with this rate
+      await expect(stRSR.connect(owner).resetStakes()).to.be.revertedWith('rate still safe')
+
+      // Add RSR to decrease stake rate - still safe
+      await rsr.connect(owner).transfer(stRSR.address, addAmt1)
+
+      // Advance to the end of noop period
+      await setNextBlockTimestamp(Number(ONE_PERIOD.add(await getLatestBlockTimestamp())))
+      await stRSR.payoutRewards()
+
+      // Calculate payout amount
+      const decayFn = makeDecayFn(await stRSR.rewardRatio())
+      const addedRSRStake = addAmt1.sub(decayFn(addAmt1, 1)) // 1 round
+      const newRate: BigNumber = fp(stakeAmt.add(addedRSRStake)).div(stakeAmt)
+
+      // Payout rewards - Advance to get 1 round of rewards
+      await setNextBlockTimestamp(Number(ONE_PERIOD.add(await getLatestBlockTimestamp())))
+      await expect(stRSR.payoutRewards()).to.emit(stRSR, 'ExchangeRateSet')
+      expect(await stRSR.exchangeRate()).to.be.closeTo(newRate, 1)
+      expect(await stRSR.totalSupply()).to.equal(stakeAmt)
+      expect(await stRSR.balanceOf(addr1.address)).to.equal(stakeAmt)
+
+      // Attempt to reset stakes, still not possible
+      await expect(stRSR.connect(owner).resetStakes()).to.be.revertedWith('rate still safe')
+
+      // Add a large amount of funds - rate will be unsafe
+      await rsr.connect(owner).mint(owner.address, addAmt2)
+      await rsr.connect(owner).transfer(stRSR.address, addAmt2)
+
+      // Advance to the end of noop period
+      await setNextBlockTimestamp(Number(ONE_PERIOD.add(await getLatestBlockTimestamp())))
+      await stRSR.payoutRewards()
+
+      // Payout rewards - Advance time - rate will be unsafe
+      await setNextBlockTimestamp(Number(ONE_PERIOD.mul(100).add(await getLatestBlockTimestamp())))
+      await expect(stRSR.payoutRewards()).to.emit(stRSR, 'ExchangeRateSet')
+      expect(await stRSR.exchangeRate()).to.be.gt(newRate) // increased
```

Comment:
Can we also add checks here that `stRSR.exchangeRate() >= fp('1e6') && stRSR.exchangeRate() <= fp('1e9')`? And same for the max case, can check that `stRSR.exchangeRate() <= fp('1e-6') && stRSR.exchangeRate() >= fp('1e-9')`.